internal/osutil/user.go

```diff
@@ -75,3 +79,40 @@ func UidGid(u *user.User) (sys.UserID, sys.GroupID, error) {
 
 	return sys.UserID(uid), sys.GroupID(gid), nil
 }
+
+// NormalizeUidGid returns the "normalized" UID and GID for the given IDs and
+// names. UID and GID take precedence over username and group. If user is
+// specified but not gid or group, the user's primary group ID is returned.
+func NormalizeUidGid(uid, gid *int, username, group string) (*int, *int, error) {
+	if uid == nil && username == "" && gid == nil && group == "" {
+		return nil, nil, nil
+	}
+	if uid == nil && username != "" {
+		u, err := userLookup(username)
+		if err != nil {
+			return nil, nil, err
+		}
+		n, _ := strconv.Atoi(u.Uid)
+		uid = &n
+		if gid == nil && group == "" {
+			// Group not specified; use user's primary group ID
+			gidVal, _ := strconv.Atoi(u.Gid)
+			gid = &gidVal
+		}
+	}
+	if gid == nil && group != "" {
+		g, err := userLookupGroup(group)
+		if err != nil {
+			return nil, nil, err
+		}
+		n, _ := strconv.Atoi(g.Gid)
+		gid = &n
+	}
+	if uid == nil && gid != nil {
+		return nil, nil, fmt.Errorf("must specify user, not just group")
+	}
+	if uid != nil && gid == nil {
+		return nil, nil, fmt.Errorf("must specify group, not just UID")
```

Comment:
It seems surprising that you could specify just 'user' and we will find the group id, but if you specify uid, we don't find the group.
Can you look up the user information by UID?

Reply:
Yeah, it's slightly odd. The rationale is I don't want to do the lookup by UID (which you can do using `user.LookupId`) because these lookups require `/etc/passwd` to be present, which may not always be the case in a container. So if you specify IDs I want to avoid the lookup.

Reply:
Interesting trade-off indeed. I can't think of a strong reason to go one way or the other right now, and I see Ben's motivation as reasonable. I suggest going forward with Ben's approach for now and observing how people use it. Once we depend on passwd, we cannot undo it.